From 86e26a3c2abaf2f7da1a340b9c73522dce40f683 Mon Sep 17 00:00:00 2001
From: Shaun Zhou <uranus-neptune@hotmail.com>
Date: Sun, 16 Jul 2023 23:24:36 +0800
Subject: [PATCH] eSPI: add Aspeed AST2500 eSPI driver to boot a host with PCH
 runs on eSPI

When PCH works under eSPI mode, the PMC (Power Management Controller) in
PCH is waiting for SUS_ACK from BMC after it alerts SUS_WARN. It is in
dead loop if no SUS_ACK assert. This is the basic requirement for the BMC
works as eSPI slave.

Also for the host power on / off actions, from BMC side, the following VW
(Virtual Wire) messages are done in firmware:
1. SLAVE_BOOT_LOAD_DONE / SLAVE_BOOT_LOAD_STATUS
2. SUS_ACK
3. OOB_RESET_ACK
4. HOST_RESET_ACK

Signed-off-by: Haiyue Wang <haiyue.wang@linux.intel.com>
---
 arch/arm/boot/dts/aspeed-g5.dtsi | 7 +++++++
 drivers/soc/aspeed/Kconfig       | 5 +++++
 drivers/soc/aspeed/Makefile      | 2 ++
 3 files changed, 14 insertions(+)

diff --git a/arch/arm/boot/dts/aspeed-g5.dtsi b/arch/arm/boot/dts/aspeed-g5.dtsi
index 04f98d1dbb97..93d71cf8acb4 100644
--- a/arch/arm/boot/dts/aspeed-g5.dtsi
+++ b/arch/arm/boot/dts/aspeed-g5.dtsi
@@ -301,6 +301,13 @@ adc: adc@1e6e9000 {
 				status = "disabled";
 			};
 
+			espi: espi@1e6ee000 {
+				compatible = "aspeed,ast2500-espi-slave";
+				reg = <0x1e6ee000 0x100>;
+				interrupts = <23>;
+				status = "disabled";
+			};
+
 			video: video@1e700000 {
 				compatible = "aspeed,ast2500-video-engine";
 				reg = <0x1e700000 0x1000>;
diff --git a/drivers/soc/aspeed/Kconfig b/drivers/soc/aspeed/Kconfig
index 3f3bf7e9db08..058a79e170e5 100644
--- a/drivers/soc/aspeed/Kconfig
+++ b/drivers/soc/aspeed/Kconfig
@@ -92,6 +92,11 @@ config ASPEED_XDMA
 	  SoCs. The XDMA engine can perform PCIe DMA operations between the BMC
 	  and a host processor.
 
+config ASPEED_ESPI_SLAVE
+	depends on ARCH_ASPEED || COMPILE_TEST
+	depends on REGMAP_MMIO
+	tristate "Aspeed ast2500 eSPI slave device driver"
+
 config ASPEED_ESPI
 	bool "ASPEED eSPI slave driver"
 	select AST2500_ESPI if MACH_ASPEED_G5
diff --git a/drivers/soc/aspeed/Makefile b/drivers/soc/aspeed/Makefile
index 39753197e272..c0d6b60f9c25 100644
--- a/drivers/soc/aspeed/Makefile
+++ b/drivers/soc/aspeed/Makefile
@@ -19,3 +19,5 @@ obj-$(CONFIG_ARCH_ASPEED)		+= aspeed-usb-ahp.o
 obj-$(CONFIG_ASPEED_MCTP)		+= aspeed-mctp.o
 obj-$(CONFIG_ASPEED_OTP)		+= aspeed-otp.o
 obj-$(CONFIG_ASPEED_SBC)		+= aspeed-sbc.o
+obj-$(CONFIG_ASPEED_ESPI_SLAVE)	+= aspeed-espi-slave.o
+
-- 
2.34.1

